test/StripeResource.spec.js

```diff
@@ -32,4 +35,245 @@ describe('StripeResource', function() {
+            amount: 1000,
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function(err, charge) {
+          expect(charge.id).to.equal('ch_123');
+          done();
+        });
+      });
+
+      it('should not retry on a 400 error', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .reply(400, {
+            error: {
+              message: 'You goofed'
+            }
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function(err) {
+          expect(err).to.not.be.null;
+          done();
+        });
+      });
+
+      it('should not retry on a 500 error', function(done) {
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .reply(500, {
+            error: {
+              message: 'We goofed'
+            }
+          });
+
+        realStripe.charges.create(options.data, function(err) {
+          expect(err).to.not.be.null;
+          done();
+        });
+      })
+
+      it('should add an idempotency key for retries using the POST method', function(done) {
+        var headers;
+
+        // Fail the first request but succeed on the 2nd.
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .reply(function(uri, requestBody, cb) {
+            headers = this.req.headers;
+
+            return cb(null, [200, {
+              id: 'ch_123"',
+              object: 'charge',
+              amount: 1000,
+            }]);
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, function() {
+          expect(headers).to.have.property('idempotency-key');
+          done();
+        });
+      });
+
+      it('should reuse the given idempotency key provided for retries', function(done) {
+        var key = uuid();
+        var headers;
+
+        nock('https://' + options.host)
+          .post(options.path, options.params)
+          .replyWithError('bad stuff')
+          .post(options.path, options.params)
+          .reply(function(uri, requestBody, cb) {
+            headers = this.req.headers;
+
+            return cb(null, [200, {
+              id: 'ch_123"',
+              object: 'charge',
+              amount: 1000,
+            }]);
+          });
+
+        realStripe.setMaxNetworkRetries(1);
+
+        realStripe.charges.create(options.data, {idempotency_key: key}, function() {
+          expect(headers['idempotency-key']).to.equal(key);
+          done();
+        });
+      });
+    });
+
+    describe('_shouldRetry', function() {
+      it('should return false if we\'ve reached maximum retries', function() {
+        var res = stripe.invoices._shouldRetry({
+          statusCode: 429
+        }, 0);
```

Comment:
Using the zero case here makes me a bit nervous; let's `stripe.setMaxNetworkRetries(1);` and use `1` here? This _could_ just be testing the case where `stripe.setMaxNetworkRetries` is unset (in theory)